apollo-federation/src/query_plan/query_planning_traversal.rs

```diff
@@ -158,6 +159,29 @@ impl BestQueryPlanInfo {
     }
 }
 
+// PORT_NOTE: This is a (partial) port of `QueryPlanningTraversal.debugStack` JS method.
+fn format_open_branch(
+    f: &mut std::fmt::Formatter<'_>,
+    data: &(&Selection, &Vec<SimultaneousPathsWithLazyIndirectPaths>),
+) -> std::fmt::Result {
+    let selection = data.0;
+    let options = data.1;
+    writeln!(f, "{selection}")?;
+    writeln!(f, " * Options:")?;
+    for option in options {
+        writeln!(f, "   - {option}")?;
+    }
+    Ok(())
+}
+
+#[allow(dead_code)] // suppress warning: tracing utility
+fn open_branch_to_string(
```

Comment:
This should probably be moved to `src/utils/logging.rs` and that whole module should probably be marked with `#![allow(dead_code)]`.